lxc/rebuild.go

```diff
@@ -0,0 +1,229 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/lxc/lxd/client"
+	"github.com/lxc/lxd/lxc/config"
+	"github.com/lxc/lxd/lxc/utils"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	cli "github.com/lxc/lxd/shared/cmd"
+	"github.com/lxc/lxd/shared/i18n"
+)
+
+// Rebuild.
+type cmdRebuild struct {
+	global    *cmdGlobal
+	flagEmpty bool
+}
+
+func (c *cmdRebuild) Command() *cobra.Command {
+	cmd := &cobra.Command{}
+	cmd.Use = usage("rebuild", i18n.G("[[<remote>:]<image>...] [<remote>:]<instance>"))
+	cmd.Short = i18n.G("Rebuild instances")
+	cmd.Long = cli.FormatSection(i18n.G("Description"), i18n.G(
+		`Rebuild instances`))
+
+	cmd.RunE = c.Run
+	cmd.Flags().BoolVar(&c.flagEmpty, "empty", false, i18n.G("Rebuild as an empty instance"))
+
+	return cmd
+}
+
+func (c *cmdRebuild) rebuild(conf *config.Config, args []string) error {
+	var name string
+	var image string
+	var remote string
+	var iremote string
+	var err error
+
+	if len(args) > 0 {
+		if len(args) == 1 {
+			remote, name, err = conf.ParseRemote(args[0])
+			if err != nil {
+				return err
+			}
+		} else if len(args) == 2 {
+			iremote, image, err = conf.ParseRemote(args[0])
+			if err != nil {
+				return err
+			}
+
+			remote, name, err = conf.ParseRemote(args[1])
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if c.flagEmpty {
+		if len(args) > 1 {
+			return fmt.Errorf(i18n.G("--empty cannot be combined with an image name"))
+		}
+
+		if len(args) == 0 {
+			return fmt.Errorf(i18n.G("--empty requires an instance name"))
+		}
+	}
+
+	d, err := conf.GetInstanceServer(remote)
+	if err != nil {
+		return err
+	}
+
+	// We are not rebuilding just a snapshot but an instance
+	if strings.Contains(name, shared.SnapshotDelimiter) {
+		return fmt.Errorf(i18n.G("Instances may not contain snapshots: %s"), name)
```

Comment:
This is a bit of a confusing error. For rebuilds we do indeed not allow instances that have snapshots to be rebuilt, but that is not what this check is checking for. Instead this is checking that the rebuild request isn't for an instance snapshot.

As such a better error IMHO would be `Instance snapshots cannot be rebuilt`.

Reply:
Also I think this check should be moved into the server side and left out of the client/CLI, as we need to check for it there too, and this avoids duplication of logic.